File: 1900.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <algorithm>
#include <stack>
using namespace std;
long long dp[501][501];
int dpr[501][501];

void showStack(int u, int d) {
    stack<int> stack;
    while(d > 0)
    {
        stack.push(u);
        u = dpr[u][d --];
    }
    while(! stack.empty())
    {
        u = stack.top();
        stack.pop();
        printf("%d ", u);
    }

//    if (d <= 0)
//    {
//        return;
//    }
//    else
//    {
//        showStack(dpr[u][d], d - 1);
//        printf("%d ", u);
//    }
}


int main() {
    int n, k;
    cin >> n >> k;
    long long a[501][501];
    long long parquet[501][501];
    long long s = 0;
    long long m = 0;
    for (int j = 0; j < 501; j ++)
        parquet[0][j] = parquet[j][0] = a[0][j] = 0;
    for (int i = 1; i < n; ++ i)
    {
        for (int j = 0; j < n - i; ++ j)
        {
            cin >> a[i][n - j - i - 1];
        }
    }
    for (int j = 0; j < n; ++ j)
    {
        s = 0;
        for (int i = 1; i < n - j; ++ i)
        {
            s += a[i][j];
            parquet[i][j] = parquet[i][j - 1] + s;
        }
    }

//    for (int i = 0; i < n; i ++)
//    {
//        for (int j = 0; j < n; j ++)
//            printf("%d ", a[i][j]);
//        printf("\n");
//    }

    for (int j = 0; j < 501; j ++)
        dp[0][j] = 0;
    for (int i = 1; i < n; i ++)
    {
        dp[i][0] = 0;
        for (int j = 1; j <= min(k,i); j ++)
        {
            m = 0;
            for (int u = 0; u < i; u ++)
            {
                s = dp[u][j - 1] + parquet[i][n - i - 1] - parquet[u][n - i - 1];
                if (s > m)
                {
                    dpr[i][j] = u;
                    m = s;
                }
            }
            dp[i][j] = m;
        }
    }
    
    long long ans = 0;
    int pos = 0;
    for (int i = 0; i <= n; i ++)
    {
        if (dp[i][k] > ans)
        {
            ans = dp[i][k];
            pos = i;
        }
    }

    cout << ans << endl;
    showStack(dpr[pos][k], k - 1);
    printf("%d\n", pos);
    //cout << dpr[pos][k];

    return 0;
}

File: 1900.py
n,k = (int(x) for x in raw_input().split(' '))

a = [[0 for i in range(n)]]
for i in range(n - 1):
    b = [int(x) for x in raw_input().split(' ')]
    b.reverse()
    a.append(b)

parquet = [[0 for j in range (n)] for i in range(n)]

for j in range(n):
    s = 0
    for i in range(1, n - j):
        s += a[i][j]
        parquet[i][j] = parquet[i][j - 1] + s

dp = [[0 for i in range(k+1)]]
dpr = [["" for i in range(k+1)] for j in range(n)]
for i in range(1, n):
    b = [0]
    for j in range(1,k+1):
        m = 0
        for u in range(i):
            s = dp[u][j - 1]
            #s += sum(a[t][k] for t in range(u + 1,i + 1) for k in range(n - i))
            #s += sua[i][u]
            s += parquet[i][n - i - 1] - parquet[u][n - i - 1]
            if s > m:
                dpr[i][j] = dpr[u][j - 1] + ' ' + str(i)
                m = s
        b.append(m)
    dp.append(b)

ans = 0
anss = ''
for i in range(n):
    if dp[i][k] > ans:
        ans = dp[i][k]
        anss = dpr[i][k]

print(ans)
print(anss[1:])

#print(m)
#print(dp)
#print(dpr)
#for i in range(n - 1):
